### PR TITLE
Fix cast nil embed

### LINKED ISSUE #19: Casting nil embed when embed already populated
Hi,

I have an issue casting nil when the embed is not empty, If i pass in a new map, it works as expected; However when the map contains values, it fails.

### PR DESCRIPTION
**Pull request by maennchen**
Fixes #19

Based on test from #20

Based on dump from #21

### CODE REVIEW

**Comment by maennchen** [comment]
@mathieuprog Can we also have a look at this?

**Comment by mathieuprog** [comment]
On it:) Be back later

**Comment by mathieuprog** [comment]
Why did you write:
```
changeset.params
|> Map.take([field, to_string(field)])
```

`:params` can contain atom keys? I think not?
https://hexdocs.pm/ecto/Ecto.Changeset.html#t:t/0

**Comment by maennchen** [comment]
@mathieuprog You're correct, I'll fix that.

**Comment by maennchen** [comment]
@mathieuprog Fixed

File: lib/polymorphic_embed.ex [hunk 1]
```diff
@@ -37,29 +37,34 @@ defmodule PolymorphicEmbed do
...
+            map_from_struct(data, :polymorphic_embed, metadata)
+          end
 
-    if data_for_field || params_for_field do
-      params =
-        Map.merge(data_for_field || %{}, params_for_field || %{})
-        |> convert_map_keys_to_string()
+        params =
+          Map.merge(data_for_field || %{}, params_for_field || %{})
+          |> convert_map_keys_to_string()
```
Comment (mathieuprog) [inline lib/polymorphic_embed.ex:57]: We should call `convert_map_keys_to_string` only for `data_for_field`, so we can move it up and pipe it under that `if`.

Comment (mathieuprog) [reply lib/polymorphic_embed.ex:57]: (that was my mistake)

File: lib/polymorphic_embed.ex [hunk 2]
```diff
@@ -37,29 +37,34 @@ defmodule PolymorphicEmbed do
...
+        changeset
+
+      {:ok, nil} ->
+        Ecto.Changeset.cast(changeset, %{to_string(field) => nil}, [field])
 
-    params_for_field = Map.get(changeset.params, to_string(field))
+      {:ok, params_for_field} ->
+        data_for_field =
+          if data = Map.fetch!(changeset.data, field) do
+            map_from_struct(data, :polymorphic_embed, metadata)
```
Comment (mathieuprog) [inline lib/polymorphic_embed.ex:52]: The pattern matching through the atom `:polymorphic_embed` is no longer necessary thanks to your previous PR.

**Review by mathieuprog — CHANGES_REQUESTED** [review]

File: lib/polymorphic_embed.ex (hunk 2 above)
Comment (maennchen) [reply lib/polymorphic_embed.ex:52]: I'll remove it

File: lib/polymorphic_embed.ex (hunk 1 above)
Comment (maennchen) [reply lib/polymorphic_embed.ex:57]: If I move that to only `data_for_field`, I get the following test errors:

<details>
<summary>ExUnit Output</summary>

```
mix test
Compiling 3 files (.ex)

15:27:53.354 [info]  == Running 20000101000000 PolymorphicEmbed.CreateTables.change/0 backward

15:27:53.356 [info]  drop table reminders

15:27:53.359 [info]  == Migrated 20000101000000 in 0.0s
warning: redefining module PolymorphicEmbed.CreateTables (current version defined in memory)
  test/support/migrations/20000101000000_create_tables.exs:1


15:27:53.398 [info]  == Running 20000101000000 PolymorphicEmbed.CreateTables.change/0 forward

15:27:53.399 [info]  create table reminders

15:27:53.406 [info]  == Migrated 20000101000000 in 0.0s


  1) test keep existing data (PolymorphicEmbedTest)
     test/polymorphic_embed_test.exs:202
     ** (Ecto.CastError) expected params to be a map with atoms or string keys, got a map with mixed keys: %{:__type__ => "sms", :number => "54", "__type__" => :sms, "attempts" => [%{date: ~U[2020-05-28 07:27:05Z], result: %{success: true}}, %{date: ~U[2020-05-28 07:27:05Z], result: %{success: true}}], "country_code" => 32, "number" => "02/807.05.53", "provider" => %{__type__: :twilio, api_key: "foo"}, "result" => %{success: true}}
     code: |> Reminder.changeset(%{
     stacktrace:
       (ecto 3.5.3) lib/ecto/changeset.ex:591: anonymous fn/3 in Ecto.Changeset.convert_params/1
       (stdlib 3.13.2) maps.erl:233: :maps.fold_1/3
       (ecto 3.5.3) lib/ecto/changeset.ex:587: Ecto.Changeset.convert_params/1
       (ecto 3.5.3) lib/ecto/changeset.ex:493: Ecto.Changeset.cast/6
       (polymorphic_embed 0.14.0) test/support/models/channel/sms.ex:26: PolymorphicEmbed.Channel.SMS.changeset/2
       (polymorphic_embed 0.14.0) lib/polymorphic_embed.ex:79: PolymorphicEmbed.cast/2
       (ecto 3.5.3) lib/ecto/changeset.ex:562: Ecto.Changeset.cast_field/8
       (ecto 3.5.3) lib/ecto/changeset.ex:523: Ecto.Changeset.process_param/7
       (elixir 1.11.2) lib/enum.ex:2181: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ecto 3.5.3) lib/ecto/changeset.ex:501: Ecto.Changeset.cast/6
       (ecto 3.5.3) lib/ecto/changeset.ex:477: Ecto.Changeset.cast/4
       (polymorphic_embed 0.14.0) test/support/models/reminder.ex:27: PolymorphicEmbed.Reminder.changeset/2
       test/polymorphic_embed_test.exs:228: (test)

...

  2) test receive embed as map of values (PolymorphicEmbedTest)
     test/polymorphic_embed_test.exs:19
     ** (Ecto.CastError) expected params to be a map with atoms or string keys, got a map with mixed keys: %{:__type__ => "sms", :attempts => [%{date: ~U[2020-05-28 07:27:05Z], result: %{success: true}}, %{date: ~U[2020-05-29 07:27:05Z], result: %{success: false}}, %{date: ~U[2020-05-30 07:27:05Z], result: %{success: true}}], :number => "02/807.05.53", :provider => %{__type__: "twilio", api_key: "foo"}, :result => %{success: true}, "__type__" => :sms, "attempts" => [], "country_code" => 1, "number" => nil, "provider" => nil, "result" => nil}
     code: |> Reminder.changeset(sms_reminder_attrs)
     stacktrace:
       (ecto 3.5.3) lib/ecto/changeset.ex:591: anonymous fn/3 in Ecto.Changeset.convert_params/1
       (stdlib 3.13.2) maps.erl:233: :maps.fold_1/3
       (ecto 3.5.3) lib/ecto/changeset.ex:587: Ecto.Changeset.convert_params/1
       (ecto 3.5.3) lib/ecto/changeset.ex:493: Ecto.Changeset.cast/6
       (polymorphic_embed 0.14.0) test/support/models/channel/sms.ex:26: PolymorphicEmbed.Channel.SMS.changeset/2
       (polymorphic_embed 0.14.0) lib/polymorphic_embed.ex:79: PolymorphicEmbed.cast/2
       (ecto 3.5.3) lib/ecto/changeset.ex:562: Ecto.Changeset.cast_field/8
       (ecto 3.5.3) lib/ecto/changeset.ex:523: Ecto.Changeset.process_param/7
       (elixir 1.11.2) lib/enum.ex:2181: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ecto 3.5.3) lib/ecto/changeset.ex:501: Ecto.Changeset.cast/6
       (ecto 3.5.3) lib/ecto/changeset.ex:477: Ecto.Changeset.cast/4
       (polymorphic_embed 0.14.0) test/support/models/reminder.ex:27: PolymorphicEmbed.Reminder.changeset/2
       test/polymorphic_embed_test.exs:50: (test)

.

  3) test inputs_for/4 (PolymorphicEmbedTest)
     test/polymorphic_embed_test.exs:387
     Assertion with == failed
     code:  assert contents == ~s"<input id=\\"reminder_channel___type__\\" name=\\"reminder[channel][__type__]\\" type=\\"hidden\\" value=\\"email\\"><input id=\\"reminder_channel_address\\" name=\\"reminder[channel][address]\\" type=\\"text\\" value=\\"a\\">"
     left:  "<input id=\"reminder_channel___type__\" name=\"reminder[channel][__type__]\" type=\"hidden\" value=\"email\"><input id=\"reminder_channel_address\" name=\"reminder[channel][address]\" type=\"text\">"
     right: "<input id=\"reminder_channel___type__\" name=\"reminder[channel][__type__]\" type=\"hidden\" value=\"email\"><input id=\"reminder_channel_address\" name=\"reminder[channel][address]\" type=\"text\" value=\"a\">"
     stacktrace:
       test/polymorphic_embed_test.exs:409: (test)

..

  4) test without __type__ (PolymorphicEmbedTest)
     test/polymorphic_embed_test.exs:103
     match (=) failed
     code:  assert {:ok, %Reminder{}} = insert_result
     left:  {:ok, %PolymorphicEmbed.Reminder{}}
     right: {
              :error,
              %Ecto.Changeset{
                action: :insert,
                changes: %{date: ~U[2020-05-28 02:57:19Z], text: "This is an Email reminder"},
                constraints: [],
                data: %PolymorphicEmbed.Reminder{__meta__: #Ecto.Schema.Metadata<:built, "reminders">, channel: nil, date: nil, id: nil, inserted_at: nil, text: nil, updated_at: nil},
                empty_values: [""],
                errors: [channel: {"is invalid", []}],
                filters: %{},
                params: %{"channel" => %{address: "john@example.com", confirmed: false, valid: true}, "date" => ~U[2020-05-28 02:57:19Z], "text" => "This is an Email reminder"},
                prepare: [],
                repo: PolymorphicEmbed.Repo,
                repo_opts: [],
                required: [:date],
                types: %{channel: {:parameterized, PolymorphicEmbed, %{metadata: [%{identify_by_fields: [], module: PolymorphicEmbed.Channel.SMS, type: "sms"}, %{identify_by_fields: ["address", "confirmed"], module: PolymorphicEmbed.Channel.Email, type: "email"}], on_type_not_found: :changeset_error}}, date: :utc_datetime, id: :id, inserted_at: :naive_datetime, text: :string, updated_at: :naive_datetime},
                valid?: false,
                validations: []
              }
            }
     stacktrace:
       test/polymorphic_embed_test.exs:119: (test)

......

Finished in 0.3 seconds
16 tests, 4 failures

Randomized with seed 453377
```

</detail>

File: lib/polymorphic_embed.ex (hunk 2 above)
Comment (maennchen) [reply lib/polymorphic_embed.ex]: changed

File: lib/polymorphic_embed.ex (hunk 1 above)
Comment (mathieuprog) [reply lib/polymorphic_embed.ex:57]: I'll have a look, thanks:) I merged as all the tests passes.

**Comment by mathieuprog** [comment]
Nice contributions! If you want to tackle another issue or PR let me know. I'll work more on this library today.

**Comment by maennchen** [comment]
@mathieuprog I quickly added #25 as well.

File: lib/polymorphic_embed.ex (hunk 1 above)
Comment (mathieuprog) [reply lib/polymorphic_embed.ex:57]: Ok the keys of `:params` are always strings, but then we fetch the embed field within `:params`, and the keys of the embed can be atoms.
So `:params` will ensure only the "first" keys are strings, but not for the nested maps inside `:params`.
No change required then.

**Pull request merged**